### PR TITLE
[Client] ray.client.connect() and ray.ClientBuilder

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?
* First steps towards moving to the `ray.client(address).connect()` API

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/__init__.py [hunk 1]
```diff
@@ -106,6 +108,8 @@ def _configure_system():
     "actor",
     "available_resources",
     "cancel",
+    "client",
+    "ClientBuilder",
```
Comment (AmeerHajAli) [inline python/ray/__init__.py:112]: can we remove this?

File: python/ray/client_builder.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+class _LocalClientBuilder(ClientBuilder):
+    pass
+
+
+def _split_address(address: str) -> Tuple[str, str]:
+    """
+    Splits address by "://" into a module string and an inner_address.
+    NOTE: inner_address may contain "://"
+    """
+    module_start_idx = address.find("://")
```
Comment (AmeerHajAli) [inline python/ray/client_builder.py:65]: if :// does not exist, this returns -1, which results in a wrong behavior. Can we raise an exception of :// is not found?

File: python/ray/ray_constants.py [hunk 3]
```diff
@@ -48,6 +48,8 @@ def env_bool(key, default):
 # we attempt to start the service running at this port.
 DEFAULT_PORT = 6379
 
+RAY_ADDRESS_ENVIRONMENT_VARIABLE = "RAY_ADDRESS"
+
```
Comment (AmeerHajAli) [inline python/ray/ray_constants.py:52]: Why do we need this? Can't we already use RAY_ADDRESS?

**Review by AmeerHajAli — COMMENTED** [review]
Left a few comments.

Can we please add a few unit tests?

File: python/ray/client_builder.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ClientInfo:
+    """
+    Basic information of the remote server for a given Ray Client connection.
+    """
+    dashboard_url: str
```
Comment (AmeerHajAli) [inline python/ray/client_builder.py:19]: this seems to be set to None?

File: python/ray/__init__.py [hunk 5]
```diff
@@ -94,6 +94,8 @@ def _configure_system():
 from ray.cross_language import java_function, java_actor_class  # noqa: E402
 from ray.runtime_context import get_runtime_context  # noqa: E402
 from ray import util  # noqa: E402
+# We import ClientBuilder so that modules can inherit from `ray.ClientBuilder`.
+from ray.client_builder import client, ClientBuilder  # noqa: E402
 
 # Replaced with the current commit when building the wheels.
 __commit__ = "{{RAY_COMMIT_SHA}}"
```
[the remark this answers is not in the document]
Comment (ijrsvt) [reply python/ray/__init__.py:101]: This gets replaced when wheels are built. This placeholder only exists for local development.

File: python/ray/__init__.py [hunk 6]
```diff
@@ -94,6 +94,8 @@ def _configure_system():
 from ray.cross_language import java_function, java_actor_class  # noqa: E402
 from ray.runtime_context import get_runtime_context  # noqa: E402
 from ray import util  # noqa: E402
+# We import ClientBuilder so that modules can inherit from `ray.ClientBuilder`.
+from ray.client_builder import client, ClientBuilder  # noqa: E402
```
[the remark this answers is not in the document]
Comment (ijrsvt) [reply python/ray/__init__.py:98]: Sure, that can work!

File: python/ray/__init__.py (hunk 1 above)
Comment (ijrsvt) [reply python/ray/__init__.py:112]: the "ClientBuilder" part yes, "client", no.

The `__all__` in python is used for `import * from ray`

File: python/ray/client_builder.py (hunk 2 above)
Comment (ijrsvt) [reply python/ray/client_builder.py:65]: I can raise an exception. The case where `find(://) == -1` should be handled before this function is invoked.

File: python/ray/client_builder.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def client(address: Optional[str] = None) -> ClientBuilder:
+    """
+    Creates a ClientBuilder based on the provided address. The address can be
+    of the following forms:
+    * None -> Connects to or creates a local cluster and connects to it.
+    * local -> Creates a new cluster locally and connects to it.
+    * IP:Port -> Connects to a Ray Client Server at the given address.
+    * module://inner_addres -> load module.ClientBuilder and pass inner_address
```
Comment (ericl) [inline python/ray/client_builder.py:90]:
```suggestion
    * module://inner_address -> load module.ClientBuilder and pass inner_address
```

File: python/ray/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ijrsvt) [reply python/ray/__init__.py:98]: Actually I prefer this being as a top level API. The reason being is that we import ClientBuilder's from **other packages** as `package.ClientBuilder`. It seems a bit confusing if our ClientBuilder class is not at the top level.

File: python/ray/client_builder.py (hunk 2 above)
Comment (ijrsvt) [reply python/ray/client_builder.py:65]: @ericl Sure! That is definitely the correct way to handle this :)

File: python/ray/__init__.py (hunk 1 above)
Comment (ijrsvt) [reply python/ray/__init__.py:112]: See my comment above.

File: python/ray/client_builder.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        return self
+
+    def connect(self) -> ClientInfo:
+        """
+        Begin a connection to the address passed in via ray.client(...).
+        """
+        client_info_tuple = ray.util.connect(
+            self.address, job_config=self.job_config)
+        return ClientInfo(
+            dashboard_url=None,
```
[the remark this answers is not in the document]
Comment (ijrsvt) [reply python/ray/client_builder.py:49]: Switched to populating it.

File: python/ray/ray_constants.py (hunk 3 above)
Comment (ijrsvt) [reply python/ray/ray_constants.py:52]: I can do that in a follow up PR, there are only 8 occurrences of it (including one place where it is already defined as a constant.

File: python/ray/client_builder.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+
+class ClientBuilder:
+    """
+    Builder for a Ray Client connection.
+    """
+
+    def __init__(self, address: Optional[str]) -> None:
+        self.address = address
+        self.job_config = JobConfig()
```
Comment (pcmoritz) [inline python/ray/client_builder.py:34]: This should be `self._job_config` (it is an implementation detail)

File: python/ray/client_builder.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        Set an environment for the session.
+        """
+        self.job_config = JobConfig(runtime_env=env)
+        return self
+
+    def connect(self) -> ClientInfo:
+        """
+        Begin a connection to the address passed in via ray.client(...).
+        """
+        client_info_tuple = ray.util.client_connect.connect(
```
Comment (pcmoritz) [inline python/ray/client_builder.py:47]: Looks like the tuple is actually a dict :P

**Review by pcmoritz — APPROVED** [review]
LGTM, let's add some unit tests before merging it (to make sure that it is testable)

File: python/ray/client_builder.py (hunk 9 above)
Comment (ijrsvt) [reply python/ray/client_builder.py:34]: Done!

File: python/ray/client_builder.py (hunk 10 above)
Comment (ijrsvt) [reply python/ray/client_builder.py:47]: renaming to `client_info_dict`

**Comment by ijrsvt** [comment]
Windows Build Failure is **not** related to this PR (failure with unrelated test `tests:test_reference_counting`)

**Comment by ijrsvt** [comment]
cc @wuisawesome this is the PR (files related to ClientBuilder)

File: python/ray/tests/test_multi_node_2.py [hunk 11]
```diff
@@ -255,14 +255,14 @@ def test_wait_for_nodes(ray_start_cluster_head):
     ],
     indirect=True)
 def test_ray_client(call_ray_start):
-    from ray.util.client import ray
-    ray.connect("localhost:20000")
+    from ray.util.client import ray as ray_client
+    ray.client("localhost:20000").connect()
```
Comment (AmeerHajAli) [inline python/ray/tests/test_multi_node_2.py:259]: is this intentional?

File: python/ray/tests/test_client_builder.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    builder = client_builder.client(address)
+    assert isinstance(builder, client_builder.ClientBuilder)
+    if address in ("local", None):
+        assert isinstance(builder, client_builder._LocalClientBuilder)
+    else:
+        assert type(builder) == client_builder.ClientBuilder
+        assert builder.address == address.replace("ray://", "")
+
+
+def test_acutal_connect(ray_start_regular_shared):
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_builder.py:37]: acutal?

File: python/ray/tests/test_client_builder.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    "address",
+    ["localhost", "1.2.3.4:1200", "ray://1.2.3.4:1200", "local", None])
+def test_client(address):
+    builder = client_builder.client(address)
+    assert isinstance(builder, client_builder.ClientBuilder)
+    if address in ("local", None):
+        assert isinstance(builder, client_builder._LocalClientBuilder)
+    else:
+        assert type(builder) == client_builder.ClientBuilder
+        assert builder.address == address.replace("ray://", "")
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_builder.py:34]: should the address remove the prefix or keep it?

**Review by AmeerHajAli — APPROVED** [review]
LGTM. a few minor comments.

File: python/ray/tests/test_client_builder.py (hunk 13 above)
Comment (ijrsvt) [reply python/ray/tests/test_client_builder.py:34]: Definitely remove!

File: python/ray/tests/test_client_builder.py (hunk 12 above)
Comment (ijrsvt) [inline python/ray/tests/test_client_builder.py:37]:
```suggestion
def test_connect_to_cluster(ray_start_regular_shared):
```

File: python/ray/tests/test_multi_node_2.py (hunk 11 above)
Comment (ijrsvt) [reply python/ray/tests/test_multi_node_2.py:259]: Yes!

File: python/ray/tests/test_client_builder.py (hunk 12 above)
Comment (ijrsvt) [reply python/ray/tests/test_client_builder.py]: Renamed to `test_acutal_connect` because this test involves actually connecting.

**Comment by ijrsvt** [comment]
Merging because the only failed test is:
```
rllib:tests/test_model_imports                                         FAILED in 3 out of 3 in 25.3s
```
This has been fixed in https://github.com/ray-project/ray/pull/15799

**Pull request merged**